File: apps/mobile/ios/ReachIOS/Sources/ReachIOS/MarketplaceModels.swift
import Foundation

public struct MarketplaceItem: Identifiable, Codable {
    public let kind: String
    public let id: String
    public let name: String
    public let description: String
    public let publisher: Publisher
    public let latestVersion: String
    public let riskLevel: String
    public let tierRequired: String
    public let requiredCapabilities: [String]
    public let sideEffectTypes: [String]

    enum CodingKeys: String, CodingKey {
        case kind, id, name, description, publisher
        case latestVersion = "latest_version"
        case riskLevel = "risk_level"
        case tierRequired = "tier_required"
        case requiredCapabilities = "required_capabilities"
        case sideEffectTypes = "side_effect_types"
    }
}

public struct Publisher: Codable {
    public let name: String
    public let verified: Bool
}

public struct CatalogPage: Codable {
    public let items: [MarketplaceItem]
    public let page: Int
    public let pageSize: Int
    public let total: Int

    enum CodingKeys: String, CodingKey {
        case items, page
        case pageSize = "page_size"
        case total
    }
}

public struct InstallIntentRequest: Codable {
    public let kind: String
    public let id: String
}

public struct InstallIntentResponse: Codable {
    public let kind: String
    public let id: String
    public let resolvedVersion: String
    public let idempotencyKey: String
    public let permissionsSummary: PermissionsSummary
    public let signature: SignatureSummary
    public let publisher: Publisher
    public let tier: TierSummary

    enum CodingKeys: String, CodingKey {
        case kind, id, signature, publisher, tier
        case resolvedVersion = "resolved_version"
        case idempotencyKey = "idempotency_key"
        case permissionsSummary = "permissions_summary"
    }
}

public struct PermissionsSummary: Codable {
    public let requiredCapabilities: [String]
    public let sideEffectTypes: [String]
    public let riskLevel: String

    enum CodingKeys: String, CodingKey {
        case requiredCapabilities = "required_capabilities"
        case sideEffectTypes = "side_effect_types"
        case riskLevel = "risk_level"
    }
}

public struct SignatureSummary: Codable {
    public let method: String
    public let keyId: String
    public let verified: Bool

    enum CodingKeys: String, CodingKey {
        case method
        case keyId = "key_id"
        case verified
    }
}

public struct TierSummary: Codable {
    public let required: String
    public let allowed: Bool
    public let current: String
}

public struct InstallRequest: Codable {
    public let kind: String
    public let id: String
    public let version: String
    public let idempotencyKey: String
    public let acceptedCapabilities: [String]
    public let acceptedRisk: Bool

    enum CodingKeys: String, CodingKey {
        case kind, id, version
        case idempotencyKey = "idempotency_key"
        case acceptedCapabilities = "accepted_capabilities"
        case acceptedRisk = "accepted_risk"
    }
}

public struct InstalledConnector: Codable {
    public let id: String
    public let pinnedVersion: String
    public let installedAt: String

    enum CodingKeys: String, CodingKey {
        case id
        case pinnedVersion = "pinned_version"
        case installedAt = "installed_at"
    }
}

File: apps/mobile/android/app/src/main/java/com/reach/mobile/data/MarketplaceModels.kt
package com.reach.mobile.data

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

@Serializable
data class MarketplaceItem(
    val kind: String,
    val id: String,
    val name: String,
    val description: String,
    val publisher: Publisher,
    val versions: List<MarketplaceVersion>,
    @SerialName("latest_version") val latestVersion: String,
    @SerialName("risk_level") val riskLevel: String,
    @SerialName("tier_required") val tierRequired: String,
    @SerialName("required_capabilities") val requiredCapabilities: List<String> = emptyList(),
    @SerialName("side_effect_types") val sideEffectTypes: List<String> = emptyList(),
    val tags: List<String> = emptyList(),
    val categories: List<String> = emptyList(),
    @SerialName("docs_url") val docsUrl: String? = null,
    @SerialName("repo_url") val repoUrl: String? = null,
    val changelog: String? = null
)

@Serializable
data class Publisher(
    val name: String,
    @SerialName("key_id") val keyId: String,
    val verified: Boolean
)

@Serializable
data class MarketplaceVersion(
    val version: String,
    @SerialName("published_at") val publishedAt: String,
    @SerialName("sha256") val sha256: String,
    val signed: Boolean,
    @SerialName("signature_key_id") val signatureKeyId: String
)

@Serializable
data class CatalogPage(
    val items: List<MarketplaceItem>,
    val total: Int,
    val page: Int,
    @SerialName("page_size") val pageSize: Int
)

@Serializable
data class InstallIntentRequest(
    val kind: String,
    val id: String,
    val version: String? = null
)

@Serializable
data class InstallIntentResponse(
    val kind: String,
    val id: String,
    @SerialName("resolved_version") val resolvedVersion: String,
    @SerialName("idempotency_key") val idempotencyKey: String,
    @SerialName("permissions_summary") val permissionsSummary: PermissionsSummary,
    val signature: SignatureSummary,
    val publisher: Publisher,
    val tier: TierSummary
)

@Serializable
data class PermissionsSummary(
    @SerialName("required_capabilities") val requiredCapabilities: List<String>,
    @SerialName("side_effect_types") val sideEffectTypes: List<String>,
    @SerialName("risk_level") val riskLevel: String
)

@Serializable
data class SignatureSummary(
    val signed: Boolean,
    val verified: Boolean,
    @SerialName("signature_key_id") val signatureKeyId: String? = null,
    val status: String
)

@Serializable
data class TierSummary(
    val required: String,
    val allowed: Boolean,
    val current: String
)

@Serializable
data class InstallRequest(
    val kind: String,
    val id: String,
    val version: String,
    @SerialName("idempotency_key") val idempotencyKey: String,
    @SerialName("accepted_capabilities") val acceptedCapabilities: List<String>,
    @SerialName("accepted_risk") val acceptedRisk: Boolean
)

@Serializable
data class InstalledConnector(
    val kind: String,
    val id: String,
    @SerialName("pinned_version") val pinnedVersion: String,
    @SerialName("verified_by") val verifiedBy: String,
    val hash: String
)
